master/buildbot/configurators/janitor.py

```diff
@@ -40,32 +40,53 @@ def now():
 
 class LogChunksJanitor(BuildStep):
     name = 'LogChunksJanitor'
-    renderables = ["logHorizon"]
+    renderables = ["logHorizon", "buildersLogHorizon"]
 
-    def __init__(self, logHorizon):
+    def __init__(self, logHorizon=None, buildersLogHorizon=[]):
         super().__init__()
         self.logHorizon = logHorizon
+        self.buildersLogHorizon = buildersLogHorizon
 
     @defer.inlineCallbacks
     def run(self):
-        older_than_timestamp = datetime2epoch(now() - self.logHorizon)
-        deleted = yield self.master.db.logs.deleteOldLogChunks(older_than_timestamp)
-        self.descriptionDone = ["deleted", str(deleted), "logchunks"]
-        return SUCCESS
+        self.descriptionDone = []
+        if self.buildersLogHorizon != []:
+            builder_timestamps = {}
+            for builder in self.buildersLogHorizon:
+                builder_timestamps[builder] = datetime2epoch(now() - self.buildersLogHorizon[builder])
+            results = yield self.master.db.logs.deleteBuilderLogs(builder_timestamps)
+            for builder, deleted in sorted(results.items()):
+                self.descriptionDone += ["deleted", str(deleted), "logchunks from", builder, '\n']
+            self.descriptionDone.pop()
+
+        if self.logHorizon is not None:
+            older_than_timestamp = datetime2epoch(now() - self.logHorizon)
+            builders = []
+            if self.buildersLogHorizon != []:
+                builders = list(self.buildersLogHorizon.keys())
+            deleted = yield self.master.db.logs.deleteOldLogChunks(older_than_timestamp, exceptions=builders)
+            if self.descriptionDone != []:
+                self.descriptionDone += ['\n']
+            self.descriptionDone += ["deleted", str(deleted), "logchunks"]
+
+        if self.logHorizon is not None or self.buildersLogHorizon != []:
```

Comment:
If this condition does not match, we will return None. Is this correct return value?

Reply:
The JanitorConfigurator class already have a condition to the case neither of logHorizon nor buildersLogHorizon is given that return None. But, since I couldn't think of all the possibilities that the function might return SUCCESS when nothing happened, I chose to code defensively.

Reply:
Alright, I will investigate and see if we can have a better return option.

Reply:
Alright, I will investigate and see if we can have a better return option.